jfr/src/main/java/io/netty/handler/jfr/JfrEventHandler.java

```diff
@@ -0,0 +1,309 @@
+                          event -> event.setLocalAddress(String.valueOf(localAddress)));
+    }
+
+    @Override
+    public void connect(
+            ChannelHandlerContext ctx,
+            SocketAddress remoteAddress, SocketAddress localAddress, ChannelPromise promise) {
+        surroundWithEvent(new ConnectEvent(), ctx, () -> ctx.connect(remoteAddress, localAddress, promise), event -> {
+            event.setRemoteAddress(String.valueOf(remoteAddress));
+            event.setLocalAddress(String.valueOf(localAddress));
+        });
+    }
+
+    @Override
+    public void disconnect(ChannelHandlerContext ctx, ChannelPromise promise) {
+        surroundWithEvent(new DisconnectEvent(), ctx, () -> ctx.disconnect(promise));
+    }
+
+    @Override
+    public void close(ChannelHandlerContext ctx, ChannelPromise promise) {
+        surroundWithEvent(new CloseEvent(), ctx, () -> ctx.close(promise));
+    }
+
+    @Override
+    public void deregister(ChannelHandlerContext ctx, ChannelPromise promise) {
+        surroundWithEvent(new DeregisterEvent(), ctx, () -> ctx.deregister(promise));
+    }
+
+// NOTE: Occurs so often, currently remove for performance reasons
+//    @Override
+//    public void channelReadComplete(ChannelHandlerContext ctx) {
+//        surroundWithEvent(new ChannelReadCompleteEvent(), ctx, ctx::fireChannelReadComplete);
+//    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) {
+        surroundWithEvent(new ChannelReadEvent(), ctx, () -> ctx.fireChannelRead(msg),
+                          event -> event.setMsg(String.valueOf(msg)));
+    }
+
+// NOTE: Occurs so often, currently remove for performance reasons
+//    @Override
+//    public void read(ChannelHandlerContext ctx) {
+//        surroundWithEvent(new ReadEvent(), ctx, ctx::read);
+//    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+        surroundWithEvent(new WriteEvent(), ctx, () -> ctx.write(msg, promise),
+                          event -> event.setMsg(String.valueOf(msg)));
+    }
+
+    @Override
+    public void channelWritabilityChanged(ChannelHandlerContext ctx) {
+        surroundWithEvent(new ChannelWritabilityChangedEvent(), ctx, ctx::fireChannelWritabilityChanged,
+                          event -> event.setWritable(ctx.channel().isWritable()));
+    }
+
+    @Override
+    public void flush(ChannelHandlerContext ctx) {
+        surroundWithEvent(new FlushEvent(), ctx, ctx::flush);
+    }
+
+    // Helper methods
+
+    private static <T extends ChannelEvent> void surroundWithEvent(T event, ChannelHandlerContext ctx,
+                                                                   Runnable operation) {
+        surroundWithEvent(event, ctx, operation, t -> {
+        });
+    }
+
+    private static <T extends ChannelEvent> void surroundWithEvent(T event, ChannelHandlerContext ctx,
```

Comment:
Putting the commit in a separate method means the stack trace will start with "surroundWithEvent" which doesn't look good. It will also add about 40-50 ns to every event with a stack trace, because JFR will need to walk one additional frame. This is about the same cost as emitting one additional event without a stack trace

Not sure if it matters in this case, but if you pass the event object to a method (so it escapes), it may make it harder for the JIT to do scalarization.

Reply:
Oh, that's true. Will see if I can come up with something else...

Reply:
Reverted this change for now